internal/packing/packing.go

```diff
@@ -201,8 +177,11 @@ func (p *Page) Alloc(width, height int) *Node {
 		panic("packing: width and height must > 0")
 	}
 
-	if !p.extendFor(width, height) {
-		return nil
+	if width < minSize {
+		width = minSize
+	}
+	if height < minSize {
+		height = minSize
```

Comment:
Do we need to move this?

Reply:
I do think so - but also I believe the previous code had a bug, even though I could find no way to trigger it.

Previous code first extended for the initial width and height, then added the root node if missing, then expanded the size to at least match the minimum, then allocated.

New code first expands the size to at least match the mininum, then creates the root node if missing (these two steps can freely be switched as they're independent from each other), then allocates and expands in one step.

So the previous code had a bug: the checked and allocated sizes differ. However, as minSize is 1, I guess this simply doesn't matter at all. If minSize were, say, 16, and one were trying to allocate 2x2, it'd check for a free 2x2 spot (and will not expand if it finds a spot for it), but then actually it could try to allocate 16x16 and fail - so it could fail even though extending would still be possible. That is why I decided to move the size extending first, so it's obviously just parameter preprocessing and one get it out of the way mentally.

If you want, though, I can do the p.root == nil block first, as those two are fully independent from each other anyway; I just found it cleaner to see the size change as parameter preprocessing and thus closer to the top of the function, but I now do see I could make the diff slightly smaller by swapping them back.

There is one other behavioral difference though, which is technically a bug but I believe it does not matter: if the requested size is bigger than the maximum allowed size by the hardware, then extendFor would have failed and it wouldn't have created the root node. With my code, it then WILL allocate the root node, then fail. My basis for considering this harmless is that on the next call with a smaller size, this root node will be used and split as normally, so the "unused" root node that could be a leftover from a failed allocation doesn't matter in practice, beyond being an unnecessarily allocated object that takes up memory.

Reply:
Hmm? I still don't understand why this doesn't work:

```go
	if p.root == nil {
		p.root = &Node{
			width:  p.width,
			height: p.height,
		}
	}

	if width < minSize {
		width = minSize
	}
	if height < minSize {
		height = minSize
	}

	return p.extendForAndAlloc(width, height)
```

> I do think so - but also I believe the previous code had a bug, even though I could find no way to trigger it.

If there is a bug, please send a PR to fix the bug.

Reply:
> With my code, it then WILL allocate the root node, then fail. My basis for considering this harmless is that on the next call with a smaller size, this root node will be used and split as normally, so the "unused" root node that could be a leftover from a failed allocation doesn't matter in practice, beyond being an unnecessarily allocated object that takes up memory.

My undertanding is that this is a side-effect of your optimization PR. Is this correct?

Reply:
Is a side effect of this PR. But as said, I do not think it matters whether this root node exists or not (and in fact, it could safely be created right at construction time).

Reply:
Order reversed as requested.

Reply:
> If there is a bug, please send a PR to fix the bug first, before doing this optimization. But I still don't understand what is the bug...

BTW, that "bug" I referenced is VERY theoretical, as it requires first changing the minSize compile time constant. Right now minSize does nothing beyond preventing to add a x * 0 rectangle which would add useless nodes, but shouldn't be happening in the first place. Thus, as there's no way to reproduce it, there's no bug right now, unless I file a PR of the sort "if you patch this line of code, this other part breaks", which doesn't seem useful.